lib/services/unified_resource.go

```diff
@@ -111,72 +121,160 @@ func (cfg *UnifiedResourceCacheConfig) CheckAndSetDefaults() error {
 	}
 }
 
 // delete removes the item by key, returns NotFound error
 // if item does not exist
-func (c *UnifiedResourceCache) delete(ctx context.Context, key []byte) error {
-	if len(key) == 0 {
-		return trace.BadParameter("missing parameter key")
-	}
-	return c.read(ctx, func(tree *btree.BTreeG[*item]) error {
-		if _, ok := tree.Delete(&item{Key: key}); !ok {
-			return trace.NotFound("key %q is not found", string(key))
+func (c *UnifiedResourceCache) delete(ctx context.Context, res types.Resource) error {
+	key := resourceKey(res)
+
+	// delete generally only sends the id, so we will fetch the actual resource from our resources
+	// map and generate our sort keys. Then we can delete from the map and all the trees at once
+	resource := c.resources[key]
+
+	nameSortKey := resourceSortKey(resource, SortByName)
+	typeSortKey := resourceSortKey(resource, SortByKind)
+
+	return c.read(ctx, func(cache *UnifiedResourceCache) error {
+		if _, ok := cache.nameTree.Delete(&item{Key: nameSortKey}); !ok {
+			return trace.NotFound("key %q is not found in unified cache name sort tree", string(nameSortKey))
+		}
+		if _, ok := cache.typeTree.Delete(&item{Key: typeSortKey}); !ok {
+			return trace.NotFound("key %q is not found in unified cache type sort tree", string(typeSortKey))
 		}
+		// delete from resource map
+		delete(c.resources, key)
 		return nil
 	})
 }
 
-func (c *UnifiedResourceCache) getRange(ctx context.Context, startKey, endKey []byte, limit int) ([]resource, error) {
-	if len(startKey) == 0 {
-		return nil, trace.BadParameter("missing parameter startKey")
+func (c *UnifiedResourceCache) getSortTree(sortField string) (*btree.BTreeG[*item], error) {
+	switch sortField {
+	case SortByName:
+		return c.nameTree, nil
+	case SortByKind:
+		return c.typeTree, nil
+	default:
+		return nil, trace.NotImplemented("sorting by %v is not supporting in unified resources", sortField)
 	}
-	if len(endKey) == 0 {
-		return nil, trace.BadParameter("missing parameter endKey")
+
+}
+
+func (c *UnifiedResourceCache) getRange(ctx context.Context, startKey []byte, matchFn func(types.ResourceWithLabels) (bool, error), req *proto.ListUnifiedResourcesRequest) ([]resource, string, error) {
+	if len(startKey) == 0 {
+		return nil, "", trace.BadParameter("missing parameter startKey")
 	}
-	if limit <= 0 {
-		limit = backend.DefaultRangeLimit
+	if req.Limit <= 0 {
+		req.Limit = backend.DefaultRangeLimit
 	}
 
 	var res []resource
-	err := c.read(ctx, func(tree *btree.BTreeG[*item]) error {
-		tree.AscendRange(&item{Key: startKey}, &item{Key: endKey}, func(item *item) bool {
-			res = append(res, item.Value)
-			if limit > 0 && len(res) >= limit {
+	var nextKey string
+	err := c.read(ctx, func(cache *UnifiedResourceCache) error {
+		tree, err := cache.getSortTree(req.SortBy.Field)
+		if err != nil {
+			return trace.Wrap(err, "getting sort tree")
+		}
+
+		iterateFunc := func(item *item) bool {
```

Comment:
I'd argue your suggestion is more readable if only for the fact that the code is read in the order it'd be executed (instead of my way which you'd see the iterate func and then it actually executes underneath it).

I was going for something like this anyway.